PSPDFViewController as an embedded Flutter view.

aksh1t opened the pull request:
Fixes #56 (iOS part)
Fixes #63 (iOS dark mode support)

# Details

Adds the following:

----

`PspdfkitView` component: The Flutter equivalent of the view component. Also adds examples on how to use it as a part of custom widgets.
- [x] iOS platform implementation
- [x] Flutter example

----

Annotation API: `addAnnotation`, `removeAnnotation`, `getAnnotations`, `getAllUnsavedAnnotations`.
- [x] iOS platform implementation
- [x] Flutter example

----

XFDF support: `importXfdf`, `exportXfdf`.
- [x] iOS platform implementation

----

Annotation Processing API: `processAnnotations`.
- [x] iOS platform implementation
- [x] Flutter example

----

# Acceptance Criteria

- [x] Before merging retest all the examples to make sure that they work on both Android and iOS.

radazzouz commented on ios/Classes/PspdfkitPlugin.m:
```diff
@@ -461,3 +462,92 @@ - (UIBarButtonItem *)barButtonItemFromString:(NSString *)barButtonItem forViewCo
+@implementation PspdfPlatformView
+
+- (nonnull UIView *)view {
+    return self.navigationController.view ? self.navigationController.view : [UIView new];
```
```suggestion
    return self.navigationController.view ?: [UIView new];
```

radazzouz commented on lib/pspdfkit_global.dart:
```diff
@@ -17,23 +17,19 @@ part 'configuration_options.dart';
   /// Sets the license key.
   static Future<void> setLicenseKey(String licenseKey) async =>
-    await _channel.invokeMethod('setLicenseKey', <String, String>{'licenseKey': licenseKey});
+      await _channel.invokeMethod('setLicenseKey', <String, String>{'licenseKey': licenseKey});
```
```suggestion
    await _channel.invokeMethod('setLicenseKey', <String, String>{'licenseKey': licenseKey});
```

radazzouz commented on ios/Classes/PspdfkitPlugin.m:
```diff
@@ -7,6 +7,8 @@
 //  This notice may not be removed from this file.
 //
 #import "PspdfkitPlugin.h"
+#import "PspdfFlutterHelper.h"
```
```suggestion
```

This causes a compile error

aksh1t replied:
Aah, my bad. I'll remove this - I was trying to clean up some code.

radazzouz commented on example/lib/main.dart:
```diff
@@ -13,15 +13,20 @@ import 'dart:async';
 const String _formPath = 'PDFs/Form_example.pdf';
 const String _instantDocumentJsonPath = 'PDFs/Instant/instant-document.json';
 const String _pspdfkitFlutterPluginTitle = 'PSPDFKit Flutter Plugin example app';
+const String _widgetExample = 'Embedded View Example';
```
```suggestion
const String _widgetExample = 'PSPDFKit Widget Example';
```

aksh1t commented:
> While testing, I noticed that there's no way out of the example. We should add the close (x) button or the back button.

I've added a back button temporarily, but I'm still not quite satisfied with the back solution. It is tricky to do it here because the home screen uses Flutter's navigation bar, but for the embedded `PSPDFViewController`, we provide a UINavigationController + nav bar from the platform side so that all the nav bar management can be done on the platform end. (tap to hide/show, customize buttons, etc).

radazzouz commented on example/lib/main.dart:
```diff
@@ -13,15 +13,20 @@ import 'dart:async';
 const String _instantDocumentJsonPath = 'PDFs/Instant/instant-document.json';
 const String _pspdfkitFlutterPluginTitle = 'PSPDFKit Flutter Plugin example app';
+const String _widgetExample = 'Embedded View Example';
+const String _widgetExampleSub = 'Opens a PDF Document as an embedded view.';
```
"Embedded view" may not be a Flutter term, I think we should use the correct Flutter terminology.
 
```suggestion
const String _widgetExampleSub = 'Opens a PDF Document in the PSPDFKit widget.';
```

aksh1t commented:
I was digging more into the memory leaks with platform views in flutter (`UiKitView`), and I think the leaks are happening on the flutter engine side. Then I read this on the `UiKitView` documentation page:

> Embedding UIViews is still in release preview... A list of open issues with embedding UIViews is available on [Github](https://github.com/flutter/flutter/issues?q=is%3Aopen+is%3Aissue+label%3A%22a%3A+platform-views%22+label%3A%22%E2%8C%BA%E2%80%AC+platform-ios%22)

There are quite a lot of memory related open issues (one that is pretty similar to ours is https://github.com/flutter/flutter/issues/45697) , so I think I'm just going to make sure we clean up everything correctly on our end and let flutter's platform view support get better.

aksh1t commented:
> In the example, rather than showing the PDF view controller full screen, I would show the PSPDFKit widget alongside another widget to illustrate that it's indeed an embedded view.

Thanks for the suggestion! I added both examples.

aksh1t commented:
@simoarpe  Could you also please review this to see if it looks okay on the iOS side?

simoarpe commented on example/lib/main.dart:
```diff
@@ -36,18 +39,44 @@ const String _formExample = 'Programmatic Form Filling Example';
 const String _importInstantJsonExample = 'Import Instant Document JSON';
 const String _importInstantJsonExampleSub = 'Shows how to programmatically import Instant Document JSON.';
+const String _widgetExampleFullScreen = 'PSPDFKit Widget Example (Fullscreen)';
+const String _widgetExampleFullScreenSub = 'Opens a PDF Document in the widget with a full screen presentation.';
```
```suggestion
const String _widgetExampleFullScreenSub = 'Opens a PDF document using the Flutter PSPDFKit widget in a full screen presentation.';
```

simoarpe commented on example/lib/main.dart:
```diff
@@ -36,18 +39,44 @@ const String _formExample = 'Programmatic Form Filling Example';
+const String _widgetExampleFullScreen = 'PSPDFKit Widget Example (Fullscreen)';
+const String _widgetExampleFullScreenSub = 'Opens a PDF Document in the widget with a full screen presentation.';
+const String _widgetExampleEmbedded = 'PSPDFKit Widget Example (Embedded)';
+const String _widgetExampleEmbeddedSub = 'Opens a PDF Document in the widget embedded with another widget.';
```
```suggestion
const String _widgetExampleEmbeddedSub = 'Opens a PDF document using the Flutter PSPDFKit widget embedded in another widget.';
```

simoarpe commented on lib/pspdfkit_global.dart:
```diff
@@ -17,23 +17,19 @@ part 'configuration_options.dart';
 /// PSPDFKit plugin to load PDF and image documents on both platform iOS and Android.
 class Pspdfkit {
-  static const MethodChannel _channel = const MethodChannel('pspdfkit');
+  static const MethodChannel _channel = const MethodChannel('com.pspdfkit.global');
```
❤️

simoarpe commented on lib/pspdfkit_widget.dart:
```diff
@@ -0,0 +1,55 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+
+class PspdfWidget extends StatefulWidget {
```
I would rename the class to `PspdfkitWidget`, it's more in line with our current naming convention and also follows the name of the file (`pspdfkit_widget.dart`).

aksh1t commented:
> I can notice some differences with the previous iOS Catalog UI, why so much space between the examples?
Shouldn't be better to have a UX that is closer to the native Catalog IU? WDYT?

I added the background so that we can see the cells in dark mode, but the spacing was not changed, it just looks that way because there is a gap between the cells, that's all. I think we can keep it as it is - if we see the system settings page in dark mode it has a similar look to this with different colored cells (although not exactly same).

> What's the difference between the embedded and the fullscreen example? Why the double entry?

That was just to demonstrate that we can use the widget in a non full screen scenario as well. But yeah both examples do seem repetitive. I removed the embedded one.

> When (and if) we'll introduce another widget I'll promote to go in favor of the same structure adopted by the dev folks at Google and add a widgets folder

I think that's a good idea nevertheless even with our single widget. I moved it.

> I don't like much the filename pspdfkit_global.dart, as it may give the wrong idea that it's a container for global settings. Considering that it's our main entry point what about main.dart?
It seems like a good practice, all the examples here use main.dart

Makes sense. Changed!

radazzouz commented on ios/Classes/PspdfPlatformView.m:
```diff
@@ -0,0 +1,92 @@
+
+    _pdfViewController = [[PSPDFViewController alloc] init];
+    [_navigationController setViewControllers:@[_pdfViewController] animated:NO];
+    _pdfViewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Back" style:UIBarButtonItemStylePlain target:self action:@selector(backButtonPressed)];
```
I would change this to:

```suggestion
    _pdfViewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[PSPDFKitGlobal imageNamed:@"x"] style:UIBarButtonItemStylePlain target:self action:@selector(backButtonPressed)];
```

I think using the `X` button image is better. Also, the "Back" text was not localized.

radazzouz replied:
Also, users should be able to hide or remove the back button. This would be especially useful when embedding the widget along side other widgets on the same screen.

radazzouz commented:
After opening the "Custom Appearance" example, all other examples that I subsequently open have the same appearance:

![recording](https://user-images.githubusercontent.com/7443038/78258470-4e5a2300-74c9-11ea-9d6a-5ea71fa5ccef.gif)

aksh1t commented:
> After opening the "Custom Appearance" example, all other examples that I subsequently open have the same appearance:

I removed the custom appearance example like we discussed and created a knowledge base article for it.

radazzouz commented on example/lib/main.dart:
```diff
@@ -234,15 +416,83 @@ class _MyAppState extends State<MyApp> {
+        Container(
+        color: Colors.grey[200],
+        padding: EdgeInsets.fromLTRB(20, 10, 20, 10),
+        child: Text("Pspdfkit Global Plugin Examples",
```
How about "Pspdfkit Global View Examples"?

radazzouz commented on example/lib/main.dart:
```diff
@@ -13,15 +13,38 @@ import 'dart:async';
+const String _formExampleForView = 'Programmatic Form Filling Example';
+const String _formExampleSubForView = 'Programmatically set and get the value of a form field using the PspdfkitView component.';
+const String _annotationsExample = 'Programmatically Add and Remove Annotations';
+const String _annotationsExampleSub = 'Programmatically add and remove annotations using the PspdfkitView component.';
```
```suggestion
const String _annotationsExampleSub = 'Programmatically add and remove annotations using the PspdfkitView widget.';
```

radazzouz commented on example/lib/main.dart:
```diff
@@ -13,15 +13,38 @@ import 'dart:async';
+const String _annotationsExample = 'Programmatically Add and Remove Annotations';
+const String _annotationsExampleSub = 'Programmatically add and remove annotations using the PspdfkitView component.';
+const String _annotationProcessingExample = 'Process Annotations';
+const String _annotationProcessingExampleSub = 'Programmatically add and remove annotations using the PspdfkitView component.';
```
```suggestion
const String _annotationProcessingExampleSub = 'Programmatically add and remove annotations using the PspdfkitView widget.';
```

aksh1t commented on example/ios/Runner/Info.plist:
```diff
@@ -45,5 +45,7 @@
 	</array>
 	<key>UIViewControllerBasedStatusBarAppearance</key>
 	<true/>
+	<key>io.flutter.embedded_views_preview</key>
```
Yeah, good point. I'll update the README too.

aksh1t commented on ios/Classes/PspdfPlatformView.m:
```diff
@@ -0,0 +1,116 @@
+@property (nonatomic) FlutterMethodChannel *channel;
+@property (nonatomic, weak) UIViewController *flutterViewController;
+@property (nonatomic) PSPDFViewController *pdfViewController;
+@property (nonatomic) UINavigationController *navigationController;
```
Hmm... true we could go with that. If I change this now, I think the appearance customization KB example that I just added today won't work since we won't be embedded in a `UINavigationController` anymore. So I will need to update that.
